### DISCUSSION TITLE
After working with Avalonia, couple of issues

### DISCUSSION THREAD

**Discussion by brunzefb** [AvaloniaUI/Avalonia#9814] [General]
Hi, I have been working to convert a WinForms app to Avalonia, so that it can run on Linux. Issues I have run into:
- FileOpenDialog broken on Linux (hangs, despite using UI thread), the exact same code works on Windows.  I have a workaround for that one.
- I can't seem to be able to resize a checkbox and make it smaller (and I don't want to replace the control template)
- I have not found a way to change the menu font size to something a bit smaller. Again, don't want to mess with the template, it should be simpler than that.
Other than that, the framework is pretty awesome, but it would have been a very steep lerning curve had I not done a lot of WPF in a very distant past. Have not run the project on my mac yet, but will do.
[Here is my UI](http://blog.brunzema.com/wp-content/uploads/2022/12/winavalonia.png)

**Comment by maxkatz6** [comment]
Hi!

> FileOpenDialog broken on Linux (hangs, despite using UI thread), the exact same code works on Windows. I have a workaround for that one.

What Avalonia version are you using? `.UseManagedDialogs()` in your app builder might help.

> I can't seem to be able to resize a checkbox and make it smaller (and I don't want to replace the control template)

Normally, you can create style to modify specific part of the template.
See:
- https://docs.avaloniaui.net/docs/styling/styles
- https://docs.avaloniaui.net/docs/styling/selectors
- https://docs.avaloniaui.net/docs/styling/troubleshooting (this page will be mostly out of dated with 11.0 previews, as some issues were solved)

> I have not found a way to change the menu font size to something a bit smaller. Again, don't want to mess with the template, it should be simpler than that.

Same as previous. But font size is inherited property (similarly how it is in WPF, not sure about WinForms). So you can write something like `<Menu TextBlock.FontSize="12" />` (`<Menu TextElement.FontSize="12" />` in 11.0)

**Reply by maxkatz6** [reply]
Checked default template for [0.10.18](https://github.com/AvaloniaUI/Avalonia/blob/0.10.18/src/Avalonia.Themes.Fluent/Controls/CheckBox.xaml). There is a problem, that Width/Height are set in XAML with local value which cannot be overridden with styles. So, in 0.10.18 you would need to redefine this template.

In 11.0-preview4 it's not an issue.

**Comment by kekekeks** [comment]
>hangs, despite using UI thread

Do not use `ShowAsync(parent).Result` and `ShowAsync(parent).GetAwaiter().GetResult()`. The code _will_ hang.

**Comment by brunzefb** [comment]
Hi
Thanks for the ideas. Tried preview, but did not work, gave me a transparent main window.

i am using 0.10.18
``` <PackageReference Include="Avalonia" Version="0.10.18" />
        <PackageReference Include="Avalonia.Desktop" Version="0.10.18" />
        <PackageReference Condition="'$(Configuration)' == 'Debug'" Include="Avalonia.Diagnostics" Version="0.10.18" />
        <PackageReference Include="Avalonia.ReactiveUI" Version="0.10.18" />
        <PackageReference Include="MessageBox.Avalonia" Version="2.1.0" />```
```
UI binds button to OpenFileCommand, then
View Model calls  View on UI thread
```
    public void OpenFileCommand()
    {
        Task.Delay(30).ContinueWith(_ => Dispatcher.UIThread.InvokeAsync(
            () =>
          { _mainWindow.OnFileOpen(); }));
    }
```
View is
```
    public void OnFileOpen()
    {
        var vm = ((App)App.Current).MainWindowViewModel;
        var hexFile = GetPath();
        vm.HexFile = string.IsNullOrEmpty(hexFile) ? vm.Prompt : hexFile;
    }
    
    private string GetPath()
    {
        var dialog = new OpenFileDialog();
        dialog.Filters.Add(new FileDialogFilter() { Name = "Hex", Extensions = { "hex" } });
        dialog.Filters.Add(new FileDialogFilter() { Name = "Bin", Extensions = { "bin" } });
        dialog.AllowMultiple = false;
        var result = dialog.ShowAsync(this).Result;
        return result?.First();
    }
```
Program.cs is
```
  private static AppBuilder BuildAvaloniaApp()
    {
        return AppBuilder.Configure<App>()
            .UsePlatformDetect()
            .UseManagedSystemDialogs()
            .LogToTrace()
            .UseReactiveUI();
    }
```
Running gets me an outline of a FileOpen Dialog,  a transparent dlg.

**Reply by maxkatz6** [reply]
>         var result = dialog.ShowAsync(this).Result;

Yeah, async operations should be awaited. Never be blocked. Especially in the crossplatform applications. On native windows you could technically open window on the separated thread and blocking wait for it, but you can't on macOS.

**Reply by maxkatz6** [reply]
> Thanks for the ideas. Tried preview, but did not work, gave me a transparent main window.

For 11.0 you would need to go through [breaking changes page](https://github.com/AvaloniaUI/Avalonia/wiki/Breaking-Changes), as default template needs to be updated (IIRC we should have update templates with version parameter now). 
But if everything works for you on stable, that's good, no need to use preview. 
I brought it up only because we had done changes for the file picker on linux.

**Comment by brunzefb** [comment]
Ok, got it to work. Was async related.
```
 public async Task OnFileOpen()
    {
        var vm = ((App)App.Current).MainWindowViewModel;
        var hexFile = await GetPath();
        vm.HexFile = string.IsNullOrEmpty(hexFile) ? vm.Prompt : hexFile;
    }
    
    private async Task<string> GetPath()
    {
        var dialog = new OpenFileDialog();
        dialog.Filters.Add(new FileDialogFilter() { Name = "Hex", Extensions = { "hex" } });
        dialog.Filters.Add(new FileDialogFilter() { Name = "Bin", Extensions = { "bin" } });
        dialog.AllowMultiple = false;
        var result = await dialog.ShowAsync(this);
        return result[0];
    }
```
Looks like the .UseManagedSystemDialogs() is needed, as it registers the needed type.

**Reply by maxkatz6** [reply]
UseManagedSystemDialogs isn't normally needed, as system file picker is used by default. But if there are problems with the system one, managed can be more reliable on linux especially.
